Fix for distributableweb, ee, remoting, hal-configuration and microprofiles tests.

jedla97 opened the pull request:
This MR include fixes for distributableweb, ee, remoting, hal-configuration and microprofiles tests modules and adding profile for EAP XP which is now mandatory for EAP. Each of the module fix is in its own commit.

In distributableweb module the css selector was edited to choose specify element. The bug was due to choosing wrong (first) element. (commit affd233)

In ee module the type setter was removed because it set wrong type in `slotModule`. (commit 63f2541)

In remoting module to create the connector the valid undertow service is needed. This fix create and remove valid `socket` and `httpListener`. (commit a183197)

In hal-configuration module the `EMPTY_SUBSYSTEMS` was updated. Some delay time for page to load up in `setUp()` method. And for pin and unpin the selector of element was changed to ensure the correct subsystem element is selected. (commit 2a58e79)  

For microprofiles module the maven profile was added for successful test when microprofiles aren't available. The `microprofile-health-api` was bumped to 2.0 due to `@Health` annotation isn't supported. With this change the deployed package was changed by adding `@Liveness` annotation. For successful deploy it's necessary to reload server which was added. In metric tests was needed to expose all subsystem and after all test the exposing of subsystem is set back to false. 
In the microprofiles is **not fully fixed** due to bug which randomly occur. (commit 9a45e22 and 190172c)

The test run for modules is available here: https://jenkins.eapqe.psi.redhat.com/job/eap-8.x-hal-3-basic-modularized-standalone/76/
Microprofiles failing but in console output is possible to see zero test run.

For microprofiles the succesful output here:
```
[INFO] Results:
[INFO] 
[WARNING] Tests run: 13, Failures: 0, Errors: 0, Skipped: 2
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] hal-testsuite 0.0.1-SNAPSHOT ....................... SUCCESS [  2.598 s]
[INFO] hal-testsuite-common ............................... SUCCESS [  6.889 s]
[INFO] resources-subsystem-fixtures ....................... SUCCESS [  3.585 s]
[INFO] tooling-deployment ................................. SUCCESS [  3.469 s]
[INFO] tests-configuration-microprofile 0.0.1-SNAPSHOT .... SUCCESS [01:18 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
```

honza-kasik commented on tests-configuration-remoting/src/test/java/org/jboss/hal/testsuite/test/configuration/remoting/HttpConnectorSecurityTest.java:
```diff
@@ -48,13 +54,18 @@
 
     @BeforeClass
     public static void beforeClass() throws Exception {
+        client.apply(new AddSocketBinding.Builder(HTTP_CONNECTOR_SECURITY).build());
```
I strongly suggest to name socket binding different (but unique name, specific to the test) than the tested connector to avoid confusion.

jedla97 replied:
Sorry for this you are absolutely right. I added specific constants for sockets and http listeners.

honza-kasik commented on tests-hal-configuration/src/test/java/org/jboss/hal/testsuite/test/configuration/SubsystemsTest.java:
```diff
@@ -56,8 +57,8 @@
     private static final Operations operations = new Operations(client);
     private static List<String> subsystems = new ArrayList<>();
     // this is from org.jboss.hal.client.configuration.subsystem.SubsystemColumn
-    private static final List<String> EMPTY_SUBSYSTEMS = asList("bean-validation", "ee-security", "jaxrs", "jdr",
```
The `jaxrs` was no longer empty?

jedla97 replied:
It's look like it. When I try EAP 7.3.9 it's look empty but 7.4.x have content in Data and Attributes and in hal/console repo jaxrs was enabled by HAL-1633. So from this two things i assume it's usable and it's not empty. 
When I think about this the test should be written for this subsystem as well.

honza-kasik reviewed (changes requested):
Thank you a lot! Really nice MR, I appreciate the description! I left a few comments behind, also, can you please revert formatting where only spaces were added?

jedla97 commented on tests-configuration-microprofile/src/test/java/org/jboss/hal/testsuite/test/configuration/microprofile/health/MicroProfileHealthCheckTest.java:
```diff
@@ -131,6 +133,7 @@ private MicroProfileHealthCheckTest openPreview() {
 
     private void doWithDeployment(Deployment deployment, Runnable action) throws Exception {
         try {
+            administration.reload();
```
Reload it's needed for successful deploy. I even tested deploy manually without reload of server but the result was that deployed package failed to deploy. With the reload the package deployed without problem and the same was result was observed in test. In short the reload here is used to setup server for deployment.

jedla97 replied:
I added condition for reload of server only when server is in `reload-required` state. I hope you meant this type of verify.

jedla97 commented on tests-configuration-microprofile/src/test/java/org/jboss/hal/testsuite/test/configuration/microprofile/metrics/MicroprofileMetricsTest.java:
```diff
@@ -13,32 +13,44 @@
 
+    @After
+    public void exposeAllSubsystems() throws Exception {
+        operations.writeListAttribute(MICROPROFILE_METRICS_ADDRESS, EXPOSED_SUBSYSTEMS, STAR);
```
I agree with you. But in case of  `@After` the state is set to expose all subsystem which is indicated by `*` in this case `STAR`. In future when this value change the `STAR` value in the tests fail the test in the same way. But what can be done is clean up the in `@AfterClass` in case the future release change the exposed system from any substytem to specific subsytem.

jedla97 replied:
I had done it. Before class the the default exposed subsystems are stored and the exposed subsystems are set to `STAR`. As I mention before when condition change this create consistency for this test.

jedla97 commented on tests-configuration-microprofile/src/test/java/org/jboss/hal/testsuite/test/configuration/microprofile/metrics/MicroprofileMetricsTest.java:
```diff
@@ -13,32 +13,44 @@
 
     private static final OnlineManagementClient client = ManagementClientProvider.createOnlineManagementClient();
     private static final Operations operations = new Operations(client);
+    private static final String STAR = "*";
```
Thanks

jedla97 commented on tests-hal-configuration/src/test/java/org/jboss/hal/testsuite/test/configuration/SubsystemsTest.java:
```diff
@@ -75,6 +76,7 @@ public static void beforeClass() throws Exception {
     @Before
     public void setUp() throws Exception {
         browser.navigate().refresh();
+        Library.letsSleep(2000);
```
It was added because in my point view the execution of code is probably fast after some tests. From this the page sometimes doesn't load and page stay blank. When you can think of some better solution just let me know.

honza-kasik replied:
Thanks for the explanation and adding the comment. Did you try either removing `browser.navigate().refresh();` or adding conditional waiting - e.g. `waitModel().until().element(By.id(Ids.ROOT_CONTAINER)).is().present();`?

jedla97 replied:
I tried go step by step with test not running on virtual screen. I find that without `browser.navigate().refresh()` the should be fine. Only needed refresh of page is after `filter` test. So sleep method was removed.

jedla97 commented:
Hi, @honza-kasik I fixed the reminders which you have or left comment there. Tested new changes where all results for fixed tests was successful [1]. 

[1] https://jenkins.eapqe.psi.redhat.com/job/eap-8.x-hal-3-basic-modularized-standalone/81/

honza-kasik reviewed (changes requested):
Thank you! Did you consider XP -> MicroProfile change we discussed on chat?

jedla97 commented:
Hi @honza-kasik ! Here are the changes you requested.  
I added fix for microprofiles for the bug which randomly occurred. The cause of this bug was that the info is saved to hashmap and the selector find 2 occurrences of same value. So it's needed to choose only first element. 

As we discus on chat I change the `xp` profile to `microprofile` profile. Regarding this `microprofile` module was deleted from basic test and added to `microprofile` profile. That mean when somebody run the test with `standalone` profile the `the microprofile` test will not run.

The successful test are here: [https://jenkins.eapqe.psi.redhat.com/job/eap-8.x-hal-3-basic-modularized-standalone/](https://jenkins.eapqe.psi.redhat.com/job/eap-8.x-hal-3-basic-modularized-standalone/)

honza-kasik reviewed (commented):
Thank you @jedla97 ! Just one comment. If you are content with it I can merge this as-is and it can be resolved later. WDYT?

honza-kasik commented:
Thank you @jedla97 ! LGTM! Merging!
